operator/src/main/resources/scripts/stopServer.sh

```diff
@@ -22,18 +22,66 @@ if [ "${MOCK_WLS}" == 'true' ]; then
   exit 0
 fi
 
-[ ! -f "${SCRIPTPATH}/wlst.sh" ]             && trace "Error: missing file '${SCRIPTPATH}/wlst.sh'."             && exit 1
+function check_for_shutdown() {
+  [ ! -f "${SCRIPTPATH}/readState.sh" ] && trace "Error: missing file '${SCRIPTPATH}/readState.sh'." && exit 1
 
-${SCRIPTDIR}/wlst.sh /weblogic-operator/scripts/stop-server.py 
+  state=`${SCRIPTPATH}/readState.sh`
+  exit_status=$?
+  if [ $exit_status -ne 0 ]; then
+    trace "Node manager not running or server instance not found; assuming shutdown" &>> /u01/oracle/stopserver.out
+    return 0
+  fi
+
+  if [ "$state" = "SHUTDOWN" ]; then
+    trace "Server is shutdown" &>> /u01/oracle/stopserver.out
+    return 0
+  fi
 
-# Return status of 2 means failed to stop a server through the NodeManager.
-# Look to see if there is a server process that can be killed.
-if [ $? -eq 2 ]; then
-  pid=$(jps -v | grep '[D]weblogic.Name=${SERVER_NAME}' | awk '{print $1}')
-  if [ ! -z $pid ]; then
-    echo "Killing the server process $pid"
-    kill -15 $pid
+  if [[ "$state" =~ ^FAILED ]]; then
+    trace "Server in failed state" &>> /u01/oracle/stopserver.out
+    return 0
   fi
+
+  trace "Server is currently in state $state" &>> /u01/oracle/stopserver.out
+  return 1
+}
+
+# Check if the server is already shutdown
+check_for_shutdown
+[ $? -eq 0 ] && trace "Server already shutdown or failed" &>> /u01/oracle/stopserver.out && exit 0
+
+# Otherwise, connect to the node manager and stop the server instance
+[ ! -f "${SCRIPTPATH}/wlst.sh" ] && trace "Error: missing file '${SCRIPTPATH}/wlst.sh'." && exit 1
+
+# Arguments for shutdown
+localAdminPort=${LOCAL_ADMIN_PORT:-${MANAGED_SERVER_PORT:-8001}}
+localAdminProtocol=${LOCAL_ADMIN_PROTOCOL:-t3}
+timeout=${SHUTDOWN_TIMEOUT:-30}
+ignoreSessions=${SHUTDOWN_IGNORE_SESSIONS:-false}
+force=${SHUTDOWN_FORCED:-true}
+
+trace "Before stop-server.py [${SERVER_NAME}] ${SCRIPTDIR}" &>> /u01/oracle/stopserver.out
+${SCRIPTPATH}/wlst.sh /weblogic-operator/scripts/stop-server.py $localAdminPort $localAdminProtocol $timeout $ignoreSessions $force &>> /u01/oracle/stopserver.out
+trace "After stop-server.py" &>> /u01/oracle/stopserver.out
+
+# at this point node manager should have terminated the server
+# but let's try looking for the server process and
+# kill the server if the process still exists,
+# just in case we failed to stop it via wlst
+pid=$(jps -v | grep "[D]weblogic.Name=${SERVER_NAME}" | awk '{print $1}')
+if [ ! -z $pid ]; then
+  echo "Killing the server process $pid" &>> /u01/oracle/stopserver.out
+  kill -15 $pid
```

Comment:
might be nice to note here that '-15' initiates forced shutdown (e.g. it's not as abrupt as a -9).